### DISCUSSION TITLE
Getting "OAuth App access restrictions" for GraphQL mutations but not same V3 action

### DISCUSSION THREAD

**Discussion by rnystrom** [community/community#24488] [Apps, API and Webhooks]
<p>Been scratching my head for a while with this one. I have an iOS app that connects to the GitHub API through a mix of V3 and GraphQL requests. I’ve been getting reports of error messages like:</p>
<pre><code>Although you appear to have the correct authorization credentials,
the `babel` organization has enabled OAuth App access restrictions, meaning that data
access to third-parties is limited. For more information on these restrictions, including
how to whitelist this app, visit
https://help.github.com/articles/restricting-access-to-your-organization-s-data/
</code></pre>
<p>If I send the following GraphQL mutation with a custom Personal Access Token with user, repo, and notifications:</p>
<pre><code>mutation {
  addReaction(input: {subjectId: "MDU6SXNzdWUzNzcxODQ3NTg=", content: HEART}) {
    subject {
      viewerCanReact
      id
    }
  }
}
</code></pre>
<p>It’ll work just fine. However, if I use a token from our app’s Basic OAuth flow, I get the FORBIDDEN error above.</p>
<p>Now where this gets funky is if I send a <a href="https://developer.github.com/v3/reactions/#create-reaction-for-an-issue" rel="nofollow noopener">V3 Reaction request</a> with the  <strong>same tokens</strong> , they both work! :confused:</p>
<p>Is there something I’m missing with authentication in our app? Or misunderstanding auth with the GraphQL API?</p>
<p><a href="https://github.com/GitHawkApp/GitHawk/issues/2165" rel="nofollow noopener">You can see all of the research we did on this issue here.</a></p>

**Comment by AndreaGriffiths11** [comment, marked as answer]
<p>HI <a class="mention" href="/rnystrom">@rnystrom</a>,</p>
<p>The GraphQL mutation works because you are using a personal access token. A personal access token is associated with a user, and a user isn’t subject to OAuth app access restrictions.</p>
<blockquote>
<p>However, if I use a token from our app’s Basic OAuth flow, I get the FORBIDDEN error above.</p>
</blockquote>
<blockquote>
<p> </p>
</blockquote>
<p>This is expected, though admittedly confusing behavior. The <code>babel</code> organization has OAuth App access restrictions enabled:</p>
<p><a href="https://help.github.com/articles/about-oauth-app-access-restrictions/" rel="nofollow noopener">https://help.github.com/articles/about-oauth-app-access-restrictions/</a></p>
<p>Because this feature is enabled, only owner-approved OAuth Apps can access the organization’s resources. In this case, Ryan’s application, GitHawk, needs to be listed as one of <code>babel</code>'s approved OAuth applications. Our documentation team wrote a guide for organization members to request an owner approve access to org resources for an OAuth App as well as the actual approval process for the owner here:</p>
<ul>
<li><a href="https://help.github.com/articles/requesting-organization-approval-for-oauth-apps/" rel="nofollow noopener">https://help.github.com/articles/requesting-organization-approval-for-oauth-apps/</a></li>
<li><a href="https://help.github.com/articles/approving-oauth-apps-for-your-organization/" rel="nofollow noopener">https://help.github.com/articles/approving-oauth-apps-for-your-organization/</a></li>
</ul>
<p>Your research on this was very insightful! If you’d like us to research further how the calls were made, please send a <code>curl&amp;nbsp; -v</code> output showcasing the full request-response pair to <a href="mailto:support@github.com">support@github.com</a> (remember to obfuscate or redact any sensitive information (authorization headers, tokens) so we can take a look!</p>
<p>Best,</p>
<p>Andrea</p>

**Comment by rnystrom** [comment]
<p>Hey <a class="mention" href="/andreagriffiths11">@andreagriffiths11</a>! Thanks for the reply. I think that this makes sense as to why the GraphQL API has auth restrictions, but that doesn’t explain why the V3 Reaction request (the exact same mutation but w/ the V3 API)  <strong>does work</strong> with the basic OAuth token.</p>
<p>If an organization has restrictions, shouldn’t both APIs behave similar for the exact same mutation?</p>
<p>Thanks!</p>

**Comment by smeijer** [comment]
<p>I’m having the exact same issue when I’m trying to browse through my private repo, with github explorer: <a href="https://developer.github.com/v4/explorer/" rel="nofollow noopener">https://developer.github.com/v4/explorer/</a></p>
<p>Is it possible to grant explorer access to my private repo?  GitHub doesn’t show up in the mentioned list in any way.</p>
<p>Never mind, I wasn’t looking. I needed to authorize trough my personal account, not through the organization. A little confusing, but makes sense if you think about it.</p>
